File: day02/main.go
package main

import (
	"io/ioutil"
	"fmt"
	"strings"
	"strconv"
)

type op func(int, int) int
type Pair struct {
	Noun int
	Verb int
}

func parseInput(data string) []int {
	input := strings.Trim(data, "\t \n")
	lines := strings.Split(input, ",")

	modules := make([]int, 0)

	for _, i := range lines {
		j, err := strconv.Atoi(i)
		if err != nil {
			panic(err)
		}

		modules = append(modules, j)
	}

	return modules
}

func execute(instructions []int, current int) {
	for {
		switch instructions[current] {
		case 1:
			handleOp(instructions, current, add)
		case 2:
			handleOp(instructions, current, mul)
		case 99:
			return
		}
		current += 4
	}
}

func handleOp(instructions []int, current int, fn op) {
	firstArg := instructions[current + 1]
	secondArg := instructions[current + 2]
	result := instructions[current + 3]

	instructions[result] = fn(instructions[firstArg], instructions[secondArg])
}

func add(a int, b int) int {
	return a + b
}

func mul(a int, b int) int {
	return a * b
}

func findPair(instructions []int, expected int) Pair {
	for i := 1; i <= 100; i++ {
		for j := 1; j <= 100; j++ {
			toRun := make([]int, len(instructions))
			copy(toRun, instructions)

			toRun[1] = i
			toRun[2] = j

			execute(toRun, 0)

			if toRun[0] == expected {
				return Pair{Noun: i, Verb: j}
			}
		}
	}

	return Pair{Noun: 0, Verb: 0}
}

func main() {
	data, _ := ioutil.ReadFile("input")
	instructions := parseInput(string(data))

	instructions[1] = 12
	instructions[2] = 2

	toRun := make([]int, len(instructions))
	copy(toRun, instructions)

	execute(toRun, 0)
	fmt.Println("first_run:", toRun[0])

	solutionPair := findPair(instructions, 19690720)
	answer := 100 * solutionPair.Noun + solutionPair.Verb
	fmt.Println("final:", answer)
}

File: day02/main.rs
fn execute(instructions: &mut Vec<usize>, current: usize) -> () {
    match instructions[current] {
        1 => handle_op(instructions, current, |a,b| a + b),
        2 => handle_op(instructions, current, |a,b| a * b),
        99 => return,
        _ => (),
    };

    execute(instructions, current + 4)
}

fn handle_op<F: Fn(usize, usize) -> usize>(instructions: &mut Vec<usize>, current: usize, op: F) {
    let first_arg = instructions[current + 1];
    let second_arg = instructions[current + 2];
    let result = instructions[current + 3];

    instructions[result] = op(instructions[first_arg], instructions[second_arg])
}

fn find_pair(instructions: &Vec<usize>, expected: usize) -> (usize, usize) {
    for i in 1..100 {
        for j in 1..100 {
            let mut to_run = instructions.to_vec();
            to_run[1] = i;
            to_run[2] = j;

            execute(&mut to_run, 0);

            if to_run[0] == expected {
                return (i, j);
            }
        }
    }

    (0, 0)
}

fn parse_instructions(input: &String) -> Vec<usize> {
    input.trim()
        .split(",")
        .map(|x| x.parse().unwrap())
        .collect()
}

fn main() {
    let input = std::fs::read_to_string("input")
        .expect("Unable to read file");

    let instructions = parse_instructions(&input);

    let mut to_run = instructions.to_vec();
    to_run[1] = 12;
    to_run[2] = 2;

    execute(&mut to_run, 0);
    println!("first_run: {}", to_run[0]);

    let pair = find_pair(&instructions, 19690720);
    let (noun, verb) = pair;
    let answer = 100 * noun + verb;

    println!("{}", answer);
}
